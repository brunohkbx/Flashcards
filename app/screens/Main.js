import React, { Component } from 'react';
import { connect } from 'react-redux'
import { View, FlatList } from 'react-native';
import Deck from '../components/Deck';
import FABContainer from '../components/FABContainer';
import FAB from '../components/FAB';
import DeckFormDialog from '../components/DeckFormDialog';
import ConfirmDialog from '../components/ConfirmDialog';
import { fetchDecks, deleteDeck } from '../actions';

export class Main extends Component {
  state = {
    formVisible: false,
    confirmRemoveDialogVisible: false,
    currentDeck: null
  }

  componentDidMount() {
    const { fetchDecks } = this.props;

    fetchDecks();
  }

  openForm = () => this.setState({ formVisible: true });
  closeForm = () => this.setState({ formVisible: false });
  openConfirmRemoveDialog = deckId => this.setState({ confirmRemoveDialogVisible: true, currentDeck: deckId });
  closeConfirmRemoveDialog = () => this.setState({ confirmRemoveDialogVisible: false, currentDeck: null });

  renderItem = ({ item }) => {
    return (
      <Deck
        title={item.title}
        flashcardsCount={item.questions.length}
        handleDeletePress={() => this.openConfirmRemoveDialog(item.id)}
      />
    );
  }

  render() {
    const {
      decks,
      deleteDeck
    } = this.props;

    const { currentDeck } = this.state;

    return (
      <View style={{flex: 1}}>
        <FlatList
          data={decks}
          renderItem={this.renderItem}
          keyExtractor={item => item.id}
        />
        <FABContainer>
          <FAB handlePress={() => this.openForm()} />
        </FABContainer>
        <DeckFormDialog
          visible={this.state.formVisible}
          handleDismiss={() => this.closeForm()}
        />
        <ConfirmDialog
          visible={this.state.confirmRemoveDialogVisible}
          handleDismiss={() => this.closeConfirmRemoveDialog()}
          handleSubmitPress={() => {this.closeConfirmRemoveDialog(); deleteDeck(currentDeck);}}
          title="Delete deck?"
          content="This deck and all it cards will be deleted. You can edit this deck if you want to change something."
          actionSubmitText="Delete"
        />
      </View>
    );
  }
}

const mapStateToProps = ({ decks }) => ({ decks: Object.values(decks) });

const mapDispatchToProps = dispatch => ({
  fetchDecks: () => dispatch(fetchDecks()),
  deleteDeck: id => dispatch(deleteDeck(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(Main);
