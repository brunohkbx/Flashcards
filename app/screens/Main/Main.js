import React, { Component } from 'react';
import { connect } from 'react-redux';
import { View, FlatList } from 'react-native';
import { Appbar, Colors } from 'react-native-paper';
import Deck from '../../components/Deck';
import ConfirmDialog from '../../components/ConfirmDialog';
import FAB from '../../components/FAB';
import Container from '../../components/Container';
import BottomRightContainer from '../../components/BottomRightContainer';
import { Movable, Fadable } from '../../components/Animations/index';
import Toaster from '../../components/Toaster';
import NoContent from './NoContent';
import SettingsDialog from './SettingsDialog';
import {
  fetchDecks,
  deleteDeck,
  loadSettings
} from '../../actions/index';
import {
  scheduleLocalNotification,
  clearLocalNotification
} from '../../lib/notifications';

export class Main extends Component {
  static navigationOptions = ({ navigation }) => {
    return {
      headerRight: (
        <Appbar.Action
          icon="settings"
          onPress={navigation.getParam('openSettingsDialog')}
          color={Colors.white}
        />
      )
    };
  };

  state = {
    dialogVisible: false,
    settingsDialogVisible: false,
    snackbarVisible: false,
    selectedDeck: {
      id: null,
      remove: false
    }
  }

  componentDidMount() {
    const { fetchDecks, navigation, loadSettings } = this.props;

    fetchDecks();
    loadSettings();

    navigation.setParams({ openSettingsDialog: this.openSettingsDialog });
  }

  componentDidUpdate(prevProps) {
    if (this.props.navigation !== prevProps.navigation) {
      const { navigation } = this.props;
      const flashMessage = navigation.getParam('flashMessage');

      if (flashMessage) {
        this.setState(
          { snackbarVisible: true },
          () => this.toaster.showMessage(flashMessage)
        );
      }
    }

    const { settings } = this.props;
    if (settings !== prevProps.settings) {
      settings.receiveNotifications ?
        scheduleLocalNotification() :
        clearLocalNotification()
    }
  }

  openDialog = deckId => {
    const selectedDeck = { ...this.state.selectedDeck, ['id']: deckId };

    this.setState({ dialogVisible: true, selectedDeck });
  }

  closeDialog = () => {
    const selectedDeck = { ...this.state.selectedDeck, ['id']: null };

    this.setState({ dialogVisible: false, selectedDeck });
  }

  submitDialog = () => {
    const selectedDeck = {...this.state.selectedDeck, ['remove']: true};

    this.setState({ dialogVisible: false, selectedDeck });
  }

  openSettingsDialog = () => this.setState({ settingsDialogVisible: true })
  closeSettingsDialog = () => this.setState({ settingsDialogVisible: false })

  addToasterRef = element => this.toaster = element;
  closeSnackbar = () => this.setState({ snackbarVisible: false });

  handleDeckDelete = id => {
    const { deleteDeck } = this.props;

    return deleteDeck(id)
      .then(() => {
        this.setState(
          { selectedDeck: { id: null, remove: false },
            snackbarVisible: true
          },
          () => this.toaster.showMessage('Deck has been successfully deleted')
        );
      });
  }

  renderItem = ({ item }) => {
    const { navigation } = this.props;
    const { selectedDeck: { id, remove }} = this.state;

    return (
      <Fadable
        fade={remove && item.id === id}
        onAnimationEnd={() => this.handleDeckDelete(item.id)}
      >
        <Deck
          title={item.title}
          flashcardsCount={item.questions.length}
          handleEditPress={
            () => navigation.navigate('EditDeck', { deckId: item.id })
          }
          handleDeletePress={() => this.openDialog(item.id)}
          onPress={() => navigation.navigate('DeckDetail', { deckId: item.id })}
        />
      </Fadable>
    );
  }

  render() {
    const { decks, navigation } = this.props;
    const {
      dialogVisible,
      selectedDeck,
      snackbarVisible,
      settingsDialogVisible
    } = this.state;

    return (
      <View style={{flex: 1}}>
        <Container>
          <FlatList
            data={decks}
            extraData={selectedDeck.remove}
            renderItem={this.renderItem}
            keyExtractor={item => item.id}
            ListEmptyComponent={NoContent}
            contentContainerStyle={[
              { flexGrow: 1 },
              decks.length ? null : { justifyContent: 'center'}
            ]}
          />

          <BottomRightContainer right={16} bottom={16} >
            <Movable toValue={snackbarVisible ? -46 : 0} >
              <FAB handlePress={() => navigation.navigate('CreateDeck')} />
            </Movable>
          </BottomRightContainer>
          <ConfirmDialog
            visible={dialogVisible}
            handleDismiss={this.closeDialog}
            handleSubmitPress={this.submitDialog}
            title="Delete deck?"
            content="This deck and all it cards will be deleted. You can edit this deck if you want to change something."
            actionSubmitText="Delete"
          />
          <SettingsDialog
            visible={settingsDialogVisible}
            onDismiss={this.closeSettingsDialog}
          />
        </Container>
        <Toaster
          ref={this.addToasterRef}
          onDismissCallback={this.closeSnackbar}
        />
      </View>
   );
  }
}

const mapStateToProps = ({ decks, settings }) => (
  {
    decks: Object.values(decks),
    settings
  }
);

const mapDispatchToProps = dispatch => ({
  fetchDecks:     () =>       dispatch(fetchDecks()),
  deleteDeck:     id =>       dispatch(deleteDeck(id)),
  loadSettings:   () =>       dispatch(loadSettings())
});

export default connect(mapStateToProps, mapDispatchToProps)(Main);
