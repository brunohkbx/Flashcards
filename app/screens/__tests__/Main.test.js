import React from 'react';
import { shallow } from 'enzyme';
import { Main } from '../Main';
import configureMockStore from 'redux-mock-store';

describe('Main', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      decks: {
        '259162c6-8c55-446b-aa4f-cf9a6fcffc2f': {
          title: 'React',
          questions: [
            {
              question: 'What is React?',
              answer: 'A library for managing user interfaces'
            }
          ],
          id: '259162c6-8c55-446b-aa4f-cf9a6fcffc2f'
        }
      },
      fetchDecks: jest.fn(),
      deleteDeck: jest.fn()
    }, propOverrides)

    const store = configureMockStore()({});

    const wrapper = shallow(<Main {...defaultProps} store={store}/>);

    const pressOnFab = () => {
      wrapper
        .find('FAB')
        .dive()
        .simulate('press')
    }

    const dismissDeckFormDialog = () => {
      wrapper
        .find('Connect(Form)')
        .dive({ context: { store }})
        .dive()
        .dive()
        .find('DeckFormDialog')
        .dive()
        .simulate('dismiss')
    }

    const dismissConfirmDialog = () => {
      wrapper
        .find('ConfirmDialog')
        .dive()
        .simulate('dismiss')
    }

    const cancelConfirmDialog = () => {
      wrapper
        .find('ConfirmDialog')
        .dive()
        .find('DialogActions')
        .childAt(0)
        .simulate('press')
    }

    const confirmConfirmDialog = () => {
      wrapper
        .find('ConfirmDialog')
        .dive()
        .find('DialogActions')
        .childAt(1)
        .simulate('press')
    }

    return {
      defaultProps,
      wrapper,
      pressOnFab,
      dismissDeckFormDialog,
      dismissConfirmDialog,
      cancelConfirmDialog,
      confirmConfirmDialog
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('fetches all the decks when component is mounted', () => {
    const mockFetchDecks = jest.fn();
    const { wrapper } = setup({ fetchDecks: mockFetchDecks });

    expect(mockFetchDecks).toHaveBeenCalled();
  });

  describe('DeckFormDialog', () => {
    it('opens when FAB is pressed', () => {
      const { wrapper, pressOnFab } = setup();
      const spy = jest.spyOn(wrapper.instance(), 'openForm');

      pressOnFab();

      expect(spy).toHaveBeenCalled();
    });

    it('closes on dismiss', () => {
      const { wrapper, dismissDeckFormDialog } = setup();
      const spy = jest.spyOn(wrapper.instance(), 'closeForm');

      dismissDeckFormDialog();

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('ConfirmDialog', () => {
    it('closes on dismiss', () => {
      const { dismissConfirmDialog, wrapper } = setup();
      const spy = jest.spyOn(wrapper.instance(), 'closeConfirmRemoveDialog');

      dismissConfirmDialog();

      expect(spy).toHaveBeenCalled();
    });

    describe('Cancel button', () => {
      it('closes', () => {
        const { cancelConfirmDialog, wrapper } = setup();
        const spy = jest.spyOn(wrapper.instance(), 'closeConfirmRemoveDialog');

        cancelConfirmDialog();

        expect(spy).toHaveBeenCalled();
      });
    });

    describe('Submit button', () => {
      it('closes', () => {
        const { confirmConfirmDialog, wrapper } = setup();
        const spy = jest.spyOn(wrapper.instance(), 'closeConfirmRemoveDialog');

        confirmConfirmDialog();

        expect(spy).toHaveBeenCalled();
      });

      it('deletes the deck', () => {
        const mockDeleteDecks = jest.fn();
        const { confirmConfirmDialog } = setup({ deleteDeck: mockDeleteDecks });

        confirmConfirmDialog();

        expect(mockDeleteDecks).toHaveBeenCalled();
      });
    });
  });
});
