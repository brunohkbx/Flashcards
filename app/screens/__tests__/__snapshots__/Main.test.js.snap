// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders properly 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <FlatList
    data={
      Object {
        "259162c6-8c55-446b-aa4f-cf9a6fcffc2f": Object {
          "id": "259162c6-8c55-446b-aa4f-cf9a6fcffc2f",
          "questions": Array [
            Object {
              "answer": "A library for managing user interfaces",
              "question": "What is React?",
            },
          ],
          "title": "React",
        },
      }
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <Styled(Component)>
    <FAB
      handlePress={[Function]}
      primary={false}
    />
  </Styled(Component)>
  <Connect(Form)
    handleDismiss={[Function]}
    visible={false}
  />
  <ConfirmDialog
    actionCancelText="cancel"
    actionSubmitText="Delete"
    content="This deck and all it cards will be deleted. You can edit this deck if you want to change something."
    handleDismiss={[Function]}
    handleSubmitPress={[Function]}
    title="Delete deck?"
    visible={false}
  />
</Component>
`;
